libcodechecker/server/client_db_access_server.py

```diff
@@ -256,6 +401,130 @@ def translate_path(self, path):
+        self.__connected = False
+
+        self.__last_connect_attempt = None
+
+    @property
+    def id(self):
+        return self.__orm_object.id
+
+    @property
+    def endpoint(self):
+        """
+        Returns the accessible URL endpoint of the product.
+        """
+        return self.__orm_object.endpoint
+
+    @property
+    def name(self):
+        """
+        Returns the display name of the product.
+        """
+        return self.__orm_object.display_name
+
+    @property
+    def session_factory(self):
+        """
+        Returns the scoped session maker on this product's database engine
+        which can be called to initiate transactional connections.
+        """
+        return self.__sc_session
+
+    @property
+    def connected(self):
+        """
+        Returns whether the product has a valid connection to the managed
+        database.
+        """
+        return self.__connected
+
+    @property
+    def last_connection_failure(self):
+        """
+        Returns the reason behind the last executed connection attempt's
+        failure.
+        """
+        return self.__last_connect_attempt[1] if self.__last_connect_attempt \
+            else None
+
+    def connect(self):
+        """
+        Initiates the actual connection to the database configured for the
+        product.
+        """
+        if self.__connected:
+            return
+
+        if self.__last_connect_attempt and \
+                (datetime.datetime.now() - self.__last_connect_attempt[0]). \
+                total_seconds() <= Product.CONNECT_RETRY_TIMEOUT:
+            return
+
+        LOG.debug("Connecting database for product '{0}'".
+                  format(self.__orm_object.endpoint))
+
+        # We need to connect to the database and perform setting up the
+        # schema.
+        LOG.debug("Configuring schema and migration...")
+        sql_server = run_database.SQLServer.from_connection_string(
+            self.__orm_object.connection,
+            self.__context.config_migration_root,
+            self.__check_env)
+
+        try:
+            sql_server.connect(self.__context.run_db_version_info, init=True)
+
+            # Create the SQLAlchemy engine.
+            LOG.debug("Connecting database engine")
+            self.__engine = sql_server.create_engine()
+            self.__sc_session = \
+                run_database.SQLServer.create_session_factory(
+                    self.__engine, sql_server.get_connection_string())
+
+            self.__connected = True
+            self.__last_connect_attempt = None
+            LOG.debug("Database connected.")
+        except Exception as ex:
+            LOG.error("The database for product '{0}' cannot be connected to."
+                      .format(self.__orm_object.endpoint))
+            LOG.error(ex.message)
+            self.__connected = False
+            self.__last_connect_attempt = (datetime.datetime.now(), ex.message)
+
+    def teardown(self):
```

Comment:
It doesn't. These database connection engines, just like the config one, are shared and alive on the entire server. These methods only exist to support loading and unloading databases mid-operations.

Contexts are created with the `scoped_session` objects instantiated from the `session_factory` property, see `do_POST` for reference.